File: python/tempmail_wrapper/providers/tempmailc.py
"""tempmailc.com provider implementation (REST API, no auth)."""

from __future__ import annotations

from datetime import datetime

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


class TempmailcProvider(TempMailProvider):
    """Provider for tempmailc.com — REST API, no auth, no cookies."""

    BASE_URL = "https://tempmailc.com/api/v1"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=False)
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "tempmailc"

    def generate_email(self) -> str:
        resp = self._http.get(f"{self.BASE_URL}/new", timeout=30)
        if resp.status_code != 200:
            raise TempMailError(f"Failed to create email: {resp.status_code}", self.name)
        data = resp.json()
        if not data.get("ok"):
            raise TempMailError("API returned not ok", self.name)
        self._email = data["email"]
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        resp = self._http.get(
            f"{self.BASE_URL}/inbox",
            params={"email": email},
            timeout=30,
        )
        if resp.status_code != 200:
            raise TempMailError(f"Inbox error: {resp.status_code}", self.name)
        data = resp.json()
        messages = []
        for item in data.get("messages", []):
            date_str = item.get("date", item.get("time", ""))
            try:
                date = datetime.fromisoformat(date_str)
            except (ValueError, AttributeError):
                date = datetime.now()
            messages.append(
                Message(
                    id=str(item.get("id", item.get("msg_id", ""))),
                    sender=item.get("from", item.get("from_mail", "")),
                    subject=item.get("subject", ""),
                    date=date,
                )
            )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        if not self._email:
            raise TempMailError("No email — call generate_email() first", self.name)
        resp = self._http.get(
            f"{self.BASE_URL}/message",
            params={"msg_id": message_id, "email": self._email},
            timeout=30,
        )
        if resp.status_code != 200:
            raise NotFoundError(f"Message {message_id} not found", self.name)
        data = resp.json()
        date_str = data.get("date", data.get("time", ""))
        try:
            date = datetime.fromisoformat(date_str)
        except (ValueError, AttributeError):
            date = datetime.now()
        return MessageDetail(
            id=str(data.get("id", message_id)),
            sender=data.get("from", data.get("from_mail", "")),
            subject=data.get("subject", ""),
            date=date,
            body_text=data.get("text", data.get("body_text", "")),
            body_html=data.get("html", data.get("body_html", "")),
            attachments=[],
        )

    def delete_email(self, email: str) -> bool:
        self._email = None
        return True

File: rust/src/providers/tempmailc.rs
use async_trait::async_trait;
use serde::Deserialize;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const BASE_URL: &str = "https://tempmailc.com/api/v1";

/// Provider for tempmailc.com — REST API, no auth, no cookies.
#[derive(Debug)]
pub struct Tempmailc {
    client: HttpClient,
    email: std::sync::Arc<tokio::sync::Mutex<Option<String>>>,
}

#[derive(Deserialize)]
struct NewResp {
    ok: bool,
    email: Option<String>,
}

impl Tempmailc {
    pub fn new(client: HttpClient) -> Self {
        Self {
            client,
            email: std::sync::Arc::new(tokio::sync::Mutex::new(None)),
        }
    }
}

fn parse_iso(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::DateTime::parse_from_rfc3339(s)
        .map(|dt| dt.with_timezone(&chrono::Utc))
        .unwrap_or_else(|_| chrono::Utc::now())
}

fn opt_str(v: Option<&serde_json::Value>) -> String {
    v.and_then(|x| x.as_str()).unwrap_or("").to_string()
}

fn opt_vid(v: Option<&serde_json::Value>) -> String {
    match v {
        Some(serde_json::Value::String(s)) => s.clone(),
        Some(serde_json::Value::Number(n)) => n.to_string(),
        Some(other) => other.to_string(),
        None => String::new(),
    }
}

#[async_trait]
impl TempMailProvider for Tempmailc {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        let resp: NewResp = self
            .client
            .get_json(&format!("{}/new", BASE_URL))
            .await?;
        if !resp.ok {
            return Err(TempMailError::Api("tempmailc API returned not ok".into()));
        }
        let email = resp
            .email
            .ok_or_else(|| TempMailError::Api("Missing email in response".into()))?;
        let mut e = self.email.lock().await;
        *e = Some(email.clone());
        Ok(email)
    }

    async fn get_inbox(&self, email: &str) -> Result<Vec<Message>, TempMailError> {
        let v: serde_json::Value = self
            .client
            .get_json(&format!("{}/inbox?email={}", BASE_URL, email))
            .await?;
        let arr = v.get("messages").and_then(|m| m.as_array());
        Ok(arr
            .map(|items| {
                items
                    .iter()
                    .map(|m| Message {
                        id: opt_vid(m.get("id").or_else(|| m.get("msg_id"))),
                        sender: opt_str(m.get("from").or_else(|| m.get("from_mail"))),
                        subject: opt_str(m.get("subject")),
                        date: parse_iso(&opt_str(m.get("date").or_else(|| m.get("time")))), preview: String::new(), has_attachments: false })
                    .collect()
            })
            .unwrap_or_default())
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        let email = self
            .email
            .lock()
            .await
            .clone()
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))?;
        let m: serde_json::Value = self
            .client
            .get_json(&format!(
                "{}/message?msg_id={}&email={}",
                BASE_URL, message_id, email
            ))
            .await?;
        Ok(MessageDetail {
            message: Message {
                id: opt_vid(m.get("id").or_else(|| m.get("msg_id"))),
                sender: opt_str(m.get("from").or_else(|| m.get("from_mail"))),
                subject: opt_str(m.get("subject")),
                date: parse_iso(&opt_str(m.get("date").or_else(|| m.get("time")))), preview: String::new(), has_attachments: false },
            body_text: {
                let b = opt_str(m.get("text").or_else(|| m.get("body_text")));
                if b.is_empty() { None } else { Some(b) }
            },
            body_html: {
                let h = opt_str(m.get("html").or_else(|| m.get("body_html")));
                if h.is_empty() { None } else { Some(h) }
            },
            attachments: vec![], ..Default::default()         })
    }

    async fn delete_email(&self, _email: &str) -> Result<bool, TempMailError> {
        let mut e = self.email.lock().await;
        *e = None;
        Ok(true)
    }
}
